gem/refactorise.py

```diff
@@ -112,6 +116,175 @@ def product(*args):
+
+        if len(atomic_index) == 0:
+            return ((), ())
+        if len(atomic_index) == 1:
+            return ((list(atomic_index.keys())[0], ), ())
+
+        # set up the ILP
+        import pulp as ilp
+        ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+        ilp_var = ilp.LpVariable.dicts('node', range(len(atomic_index)), 0, 1, ilp.LpBinary)
+
+        # Objective function
+        # Minimise number of factors to pull. If same number, favour factor with larger extent
+        big = 10000000  # some arbitrary big number
+        ilp_prob += ilp.lpSum(ilp_var[index] * (big - self.argument_indices_extent(atomic)) for atomic, index in iteritems(atomic_index))
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[index] for index in connection) >= 1
+
+        ilp_prob.solve()
+        if ilp_prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 1]
+        other_atomics = [atomic for atomic, _index in iteritems(atomic_index) if ilp_var[_index].value() == 0]
+        optimal_atomics = sorted(optimal_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        other_atomics = sorted(other_atomics, key=lambda x: self.argument_indices_extent(x), reverse=True)
+        return (tuple(optimal_atomics), tuple(other_atomics))
+
+    def factorise_atomics(self, optimal_atomics):
+        """
+        Group and factorise monomials based on a list of atomics. Create new monomials for
+        each group and optimise them recursively.
+        :param optimal_atomics: list of tuples of optimal atomics and their sum indices
+        :return: new MonomialSum object with atomics factorised.
+        """
+        if not optimal_atomics:
+            return self
+        if len(self.ordering) < 2:
+            return self
+        new_monomial_sum = MonomialSum()
+        new_monomial_sum.argument_indices = self.argument_indices
+        # Group monomials according to each optimal atomic
+        factor_group = OrderedDict()
+        for key, (_sum_indices, _atomics) in iteritems(self.ordering):
+            for (sum_indices_set, sum_indices), oa in optimal_atomics:
+                if key[0] == sum_indices_set and oa in _atomics:
+                    # Add monomial key to the list of corresponding optimal atomic
+                    factor_group.setdefault(((sum_indices_set, sum_indices), oa), []).append(key)
+                    break
+            else:
+                # Add monomials that do no have argument factors to new MonomialSum
+                new_monomial_sum.add(_sum_indices, _atomics, self.monomials[key])
+        # We should not drop monomials
+        assert sum([len(x) for x in itervalues(factor_group)]) + len(new_monomial_sum.ordering) == len(self.ordering)
```

Comment:
Why cast the generator to a list?  Write `sum(map(len, itervalues(factor_group)))`